### DISCUSSION TITLE
How to embed a Digital signature created using my own PKCS function, into the original pdf file ?

### DISCUSSION THREAD

**Discussion by dynstat** [MatthiasValvekens/pyHanko#230] [Q&A]
In my case, I already have the signature received after performing the PKCS sign operation related to the Hardware token, on the pdf.
Since I need a signed pdf now, how can I add or embed that signature to the original pdf ?

**Comment by MatthiasValvekens** [comment]
Hi @dynstat,

Sorry, I don't understand your question. Are you saying that you already computed a signature over the entire contents of the original PDF, and you now want to embed that signature into the file?

With PDF signatures as they exist today, that's not possible. Embedding a signature requires first preparing the document in a particular way, then you compute a digest over the prepared document (minus one region in the file), and then you set up a CMS signature using that digest. In other words, the "embedding" step and the "signing" step are very tightly coupled (for better or worse). You unfortunately can't just compute the signature first and then start embedding it.

That said, navigating this convoluted process for you is exactly what pyHanko does (at its core).

**Reply by dynstat** [reply]
Hi @MatthiasValvekens, Thank You for the response.  
Yes,  that's what I was trying to do. I computed a 256 bytes signature by passing the entire pdf file data to a PKCS function that computes its hash internally and returns the signature.

I found your response to a bit similar problem here  that I'm trying https://github.com/MatthiasValvekens/pyHanko/discussions/126#discussioncomment-3111679 

Is there any way that I could use/call my function just to compute the signature along with pyhanko's pdf signing functionality ?
If not, What could be the right or better approach ? 
Basically I just want to sign a pdf using a hardware token.

**Reply by MatthiasValvekens** [reply]
Yes, by extending the [Signer](https://pyhanko.readthedocs.io/en/latest/lib-guide/signing.html#extending-signer) class :)
